bgpd/bgp_mplsvpn.c

```diff
@@ -1336,6 +1336,231 @@ leak_update(struct bgp *to_bgp, struct bgp_dest *bn,
+						    struct bgp *to_bgp,
+						    struct bgp *from_bgp,
+						    afi_t afi, safi_t safi)
+{
+	struct bgp_nexthop_cache *bnc;
+	struct bgp_label_per_nexthop_cache *blnc;
+	struct bgp_label_per_nexthop_cache_head *tree;
+	struct prefix *nh_to_use_ptr;
+	struct prefix nh_from_bnc = {0};
+	struct bgp *bgp_nexthop;
+	afi_t nh_afi;
+
+	/* check the next-hop reachability in the
+	 * bgp instance where the next-hop of the entry is
+	 */
+	if (pi->extra && pi->extra->bgp_orig)
+		bgp_nexthop = pi->extra->bgp_orig;
+	else
+		bgp_nexthop = from_bgp;
+
+	nh_afi = BGP_ATTR_NH_AFI(afi, pi->attr);
+
+	if (bgp_find_or_add_nexthop(from_bgp, bgp_nexthop, nh_afi, safi, pi,
+				    NULL, 0, NULL) == 0)
+		/* no valid next-hop, wait for next-hop validity */
+		return MPLS_INVALID_LABEL;
+
+	bnc = pi->nexthop;
+	if (!bnc) {
+		/* no valid next-hop, then per-vrf label is chosen
+		 * note that the case nexthop_num == 0 should not happen
+		 * as the 'bgp_process_nexthop_update' function frees the bnc
+		 * entry
+		 */
+		return from_bgp->vpn_policy[afi].tovpn_label;
+	}
+
+	/* check that the number of available nexthop is 1
+	 */
+	if (bnc->nexthop_num > 1 ||
+	    bnc->nexthop->type == NEXTHOP_TYPE_BLACKHOLE) {
+		/* ECMP and blachole routes do not support
+		 * per-nexthop label mode;
+		 * fallback to the per-vrf label.
+		 */
+		bgp_mplsvpn_path_nh_label_unlink(pi);
+		return from_bgp->vpn_policy[afi].tovpn_label;
+	}
+
+	/* extract the resolved nexthop given by the BNC nexthop cache
+	 * if nexthop type is not connected, then use the resolved nexthop
+	 */
+	if (bnc->nexthop->type != NEXTHOP_TYPE_IFINDEX) {
+		if (bnc->nexthop->type == NEXTHOP_TYPE_IPV4 ||
+		    bnc->nexthop->type == NEXTHOP_TYPE_IPV4_IFINDEX) {
+			nh_from_bnc.family = AF_INET;
+			nh_from_bnc.prefixlen = IPV4_MAX_BITLEN;
+			nh_from_bnc.u.prefix4 = bnc->nexthop->gate.ipv4;
+		} else {
+			nh_from_bnc.family = AF_INET6;
+			nh_from_bnc.prefixlen = IPV6_MAX_BITLEN;
+			memcpy(&nh_from_bnc.u.prefix6, &bnc->nexthop->gate.ipv6,
+			       sizeof(struct in6_addr));
+		}
+		nh_to_use_ptr = &nh_from_bnc;
+	} else
+		/* the nexthop is in bnc->prefix */
+		nh_to_use_ptr = &bnc->prefix;
+
+	/* at least one bnc is connected.
+	 * allocate a nexthop label cache entry
+	 */
+	tree = &from_bgp->mpls_labels_per_nexthop[family2afi(
+		nh_to_use_ptr->family)];
+	blnc = bgp_label_per_nexthop_find(tree, nh_to_use_ptr);
+	if (!blnc) {
+		blnc = bgp_label_per_nexthop_new(tree, nh_to_use_ptr);
+		blnc->to_bgp = to_bgp;
+	}
+
+	if (pi->label_nexthop_cache != blnc) {
+		/* Unlink from existing nexthop cache, if any.
+		 * This will also free the nexthop label cache entry
+		 */
+		bgp_mplsvpn_path_nh_label_unlink(pi);
+		if (blnc) {
+			/* updates NHT pi list reference */
+			LIST_INSERT_HEAD(&(blnc->paths), pi, label_nh_thread);
+			pi->label_nexthop_cache = blnc;
+			pi->label_nexthop_cache->path_count++;
+		}
+	}
+
+	if (blnc->label == MPLS_INVALID_LABEL)
+		/* request a label to zebra for this nexthop
```

Comment:
it seems we can request the label from zebra several times. Don't do it if it was already requested

Reply:
there is a protection in the bgp_lp_get() API.